os/pm/pm_idle.c

```diff
@@ -183,6 +180,136 @@ static inline void pm_changeall(enum pm_state_e newstate)
  * Public Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: pm_idle
+ *
+ * Description:
+ *   This function is called by IDLE thread to check if board can go to sleep
+ *   or not. This function also allow to set wake up timer & handler and do all
+ *   the PM pre processing required before going to sleep.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   True  - If board can go to sleep.
+ *   False - If board can not go to sleep.
+ *
+ ****************************************************************************/
+
+bool pm_idle(void)
+{
+	irqstate_t flags;
+	enum pm_state_e newstate;
+
+	/* Decide, which power saving level can be obtained */
+	newstate = pm_checkstate();
+	/* Perform board-specific, state-dependent logic here */
+	pmvdbg("newstate= %d\n", newstate);
+	/* Then force the global state change */
+	if (pm_changestate(newstate) < 0) {
+		/* The new state change failed */
+		pmdbg("State change failed! newstate = %d\n", newstate);
+		return false;
+	}
+	/* If current state is not good to go sleep, then return false to IDLE thread */
+	if (g_pmglobals.state != PM_SLEEP) {
+		return false;
+	}
+	/* Disable interrupts through out this operation to set wakeup timer */
+	flags = enter_critical_section();
+#ifdef CONFIG_PM_TIMEDWAKEUP
+	clock_t delay = wd_getwakeupdelay();
+	if (delay > CONFIG_PM_MIN_SLEEP_TIME) {
+		pmvdbg("Setting timer and Board will wake up after %d millisecond\n", delay);
+		up_set_pm_timer(TICK2USEC(delay));
+	}
+#endif
```

Comment:
if the delay is less than the PM_MIN_SLEEP_TIME , then board should not go to sleep , so please return false here